### PR TITLE
Improve evalscript performance

### PR DESCRIPTION
**Pull request by Ardweaden**
Closes https://git.sinergise.com/team-6/openeo-platform/-/issues/112

The code produced by the converter was not optimised for speed, as it was more or less a proof of concept and thus the ease of implementation was the primary concern.

This PR improves the speed of the evalscript execution up to approximately an order of magnitude, imo making the execution time _bearable_.

Calling `_iter` was very slow and particularly `reduceByDimension` caused the biggest slowdown.

Changes with biggest effects include (for the specific script I was testing on):
- replacing `_iter` and array-inside-array data with `ndarray` library (from `1min13s` to `48s`)
- replacing `newData.push(...entry)` with `newData.concat(Object.values(entry))` in `makeArrayFromSamples`, introduced when changing to support `ndarray`. (from `48s` to `31s`). Replaced spread operators everywhere.
- replacing `JSON.parse(JSON.stringify(this.dimensions))` for copying dimensions with a for loop. (`31s` to `10s`).
- other minor changes

I added a simple script that runs a generated evalscript, instructions are in the readme. By default it runs 1 000 000 iterations, simulating processing a 1000x1000px image. The actual times can vary quite a bit depending on the computer and the load.

Old performance (current `master`):
```
1000000 iterations per test (1000x1000)

test_graph_1                            : 50.460s
reduce_mean_one_band                    : 1:20.089 (m:ss.mmm)
reduce_mean_one_band                    : 1:17.199 (m:ss.mmm)
reduce_mean_one_band                    : 1:07.611 (m:ss.mmm)
test_short_graph                        : 15.879s
test_apply_absolute                     : 37.084s
test_apply_math                         : 25.813s
test_apply_linear_scale_range           : 24.855s
test_count_with_condition               : 38.004s
test_count_without_condition            : 34.294s
```

New perfomance (this branch):
```
1000000 iterations per test (1000x1000)

test_graph_1                            : 6.056s
reduce_mean_one_band                    : 6.074s
reduce_mean_one_band                    : 6.484s
reduce_mean_one_band                    : 6.622s
test_short_graph                        : 2.884s
test_apply_absolute                     : 5.254s
test_apply_math                         : 6.545s
test_apply_linear_scale_range           : 4.907s
test_count_with_condition               : 10.007s
test_count_without_condition            : 7.547s
```

There are of course possible improvements, these were just low hanging fruits. Hardcoded specialised evalscripts are 1-2 orders of magnitude faster than the general ones. So we could handle some common cases separately. For example, many process graphs will likely be linear (doesn't branch out, so e.g. cloning the `DataCube` - which is very expensive - can be omitted), and they will likely never work with more than 2 dimensions (or 4, meaning just that spatial dimensions, temporal dimension and bands dimension, so a lot of complication with generalised handling of ndarrays can be simplified).

Feel free to suggest or implement any other ideas for improving the performance.

-------------------------------
Also fixes some incorrect tests: `add_dimension`, `rename_labels`, `filter_bands` which had invalid input `samples` (band values were arrays).

Skips `merge_cubes` tests as it wasn't properly implemented to begin with (there is a ticket for that).

Also fixes a problem with bands to hold dimension information being added to `updateOutput` even though there encoding was disabled.

### CODE REVIEW

**Review by dgostencnik — APPROVED** [review]
Comments are again purely cosmetical so fftr. 
You probably also need a computer upgrade as they  are ( also on the master)   much faster on mine 
```
this branch
test_graph_1                            : 2.064s
reduce_mean_one_band                    : 2.066s
reduce_mean_one_band                    : 2.098s
reduce_mean_one_band                    : 2.116s
test_short_graph                        : 943.178ms
test_apply_absolute                     : 1.585s
test_apply_math                         : 1.897s
test_apply_linear_scale_range           : 1.427s
test_count_with_condition               : 2.848s
test_count_without_condition            : 2.223s
```

Overall, it is quite difficult to understand what is going in DataCube.js, I got a bit frustrated and gave up.

**Comment by Ardweaden** [comment]
> Comments are again purely cosmetical so fftr. You probably also need a computer upgrade as they are ( also on the master) much faster on mine

Haha it has indeed been struggling recently, sometime running eobrowser locally slowed it down to a crawl 😆.

Anyway, the difference can also be checked in EOB. Here are the evalscript generated from `reduce_mean_one_band`. I have only changed the `sampleType` to `AUTO` and set output bands to 1 in `updateOutput`. I set the timespan to 1st to 14th march.

Old: https://sentinelshare.page.link/L5oz
New: https://sentinelshare.page.link/62xS

For normal 512x512 tiles the old ones load in 13-15s for me and the new ones in 3-5s. Analytical download 1250x933px of the same location (zoom=12&lat=43.02109&lng=11.73855) took 20s vs 1.1min.

File: src/pg_to_evalscript/javascript_datacube/DataCube.js [hunk 1]
```diff
@@ -71,14 +82,21 @@ class DataCube {
     }
 
     clone() {
-        const copy = new DataCube(this._iter(this.data), this.bands_dimension_name, this.temporal_dimension_name)
-        copy.dimensions = JSON.parse(JSON.stringify(this.dimensions))
+        const copy = new DataCube(ndarray(this.data.data.slice(), this.data.shape), this.bands_dimension_name, this.temporal_dimension_name)
+        const newDimensions = []
+        for (let dim of this.dimensions) {
```
[the remark this answers is not in the document]
Comment (Ardweaden) [reply src/pg_to_evalscript/javascript_datacube/DataCube.js:91]: No, we are making a copy of the `DataCube` instance here and one can add and remove dimensions after instantiating it. They can also change dimension names and labels.

File: src/pg_to_evalscript/javascript_datacube/DataCube.js [hunk 2]
```diff
@@ -208,17 +190,57 @@ class DataCube {
...
+                newData.push(process({
+                    "x": this.data.data[i]
+                }))
+            }
+            this.data.data = newData
+        } else {
+            const shape = this.data.shape
+            const cumulatives = fill(shape.slice(), 0);
+            const coords = shape.slice();
+            const arr = []
```
[the remark this answers is not in the document]
Comment (Ardweaden) [reply src/pg_to_evalscript/javascript_datacube/DataCube.js]: Removed.

**Pull request merged**